AI Camera Module

SakuraTheBabyfur opened the pull request:
Changes AI camera list into security camera console menu

Spacemanspark commented:
Re-posting my comment here.
"I'm going with a solid :-1:
The security monitor is obnoxious, and lags down the entire game when I open it (not sure about anyone else).
With the current setup, I can press a single key and find what I'm looking for in seconds without having to ctrl + f (Which stops working half the time when the list updates) and what-have-you.
That, and it doesn't take up the entire screen like the security monitor does."

Crazylemon64 commented on code/modules/nano/modules/camera_monitor.dm:
```diff
@@ -0,0 +1,188 @@
+
+// Switching to cameras
+/datum/nano_module/camera_monitor/proc/switch_to_camera(var/mob/user, var/obj/machinery/camera/C)
+	if ((get_dist(user, src) > 1 || user.machine != src || user.blinded || !( user.canmove ) || !( C.can_use() )) && (!istype(user, /mob/living/silicon/ai)))
```
The last clause of this line - wouldn't it be better suited for an `isAI()` check?

SakuraTheBabyfur replied:
Il play around with it tomorrow after classes, its late right now and i need some sleep.

Crazylemon64 commented on code/modules/nano/modules/camera_monitor.dm:
```diff
@@ -0,0 +1,188 @@
+		return 1
+
+	if(href_list["switchTo"])
+		if(usr.stat || ((get_dist(usr, src) > 1 || !( usr.canmove ) || usr.blinded) && !istype(usr, /mob/living/silicon))) return
```
How does get_dist work with a datum argument?

PsiOmegaDelta replied:
Not sure how close Bay and Paradise NanoUI is these days but rather than ```src``` you probably want to use ```nano_host()``` in this case.

Crazylemon64 commented:
Eep! It's a `:master` branch!

Otherwise, I don't have much to say about the contents of the branch itself, as I'm not much used to the camera monitors - anything I DO have to say, I'll comment within the file.

SakuraTheBabyfur commented:
I have never seen any negative performance with the camera monitor versus the camera list. Il play around with stuff tomorrow, maybe i can get this more optimized then the normal sec monitor, since i can basically remove all the checks since the AI is the only one that can access this module, and maybe that will make it faster.

SakuraTheBabyfur commented:
also, damn it github

Spacemanspark commented:
My other point of usability would still stand.

PsiOmegaDelta commented:
Also, is this a nearly straight copy-paste from the security camera console code?

SakuraTheBabyfur commented:
its very similar to the sec camera console, but not totally, and im planning on making some changes tomorrow, but tonight i kinda just threw it together, plus its my first time coding in byond so hopefully i will get better.

mrturkeytoe commented:
I honestly prefer it the way it is now, sure it isn't very pretty but it's fast, effective and efficient.

Aurorablade commented:
my opnion is this actually NERFs the AI camera. They already have a means to book mark and jump to bookmarked locations, plus more mobility with AI Eye movment. Unless you include both systems and make this a sort of fancy jump to list.

Fox-McCloud commented:
Not a fan :-1: This method is slow and klunky...and I've only ever played AI about a couple dozen times.

tigercat2000 commented:
@SakuraTheBabyfur Heeey, quick thing, we don't actually all hate you- AI functions are always a bit touchy, it seems. Always happy to see new coders.

SakuraTheBabyfur commented:
Before we scrap this idea completely i want to see if i can get this optimized or tweaked in some other way, or even just add this under "AI commands" and leave the list or something. I much prefer the camera monitor one as it gives you the option to choose different networks, as well as allows you to easily go between multiple cameras since the window stays open. When I tested this it was pretty fast, but i also tested it on an empty test server so who knows

PsiOmegaDelta commented:
What mainly affects how unpleasant the camera monitor is to use is the available bandwidth you have, which rarely is an issue locally or with a decent enough connection.
I have long considered making it so that the camera console only shows a single camera network at a time, and allow the user to switch, which should cut down the data send to a fair degree (as well as make it easier to find the camera you want).

BiberDark commented:
As a regular Ai player i use the camera monitor on a regular basis.  It is the only good and fast way to locate broken cams (grey ones) on the station for me and my cyborgs. Cams which are already broken at roundstart are not listed in the alarms, So it is very difficult to find them without this tool.

PS I have no lag when i use them.

SakuraTheBabyfur commented:
Alright sorry it took so long to do this (im in college) but ive made a bunch of changes. Cleaned up the code for camera module, optimized it quite a bit, so it should no longer lag (it was auto updating every master tick which honestly is not necessary). I also reduced the initial size of the pop up from 900,800 to 650,550, and keep in mind that this can be re-sized like any other window, I just thought this was a better starting size. Keep in mind those changes are only for the AI's camera monitor. Something I changed for both the AI's camera monitor and the security camera monitor however, is I moved the "Networks" to the top instead of the bottom. This means that they stay in one place instead of getting moved around whenever you activate another network, which in my opinion, is a lot cleaner. This is what it looks like now: https://i.gyazo.com/8430c80bd734c5cf159cfc9a9b0bc32e.png (let me know your guys thoughts on that)

Crazylemon64 commented:
Unfortunately, since this PR is using master as its base branch, I'm having trouble acquiring a discrete copy of the changes you've made here to actually test this with. I don't see any other major problems codewise, at least.

SakuraTheBabyfur commented:
try this then https://github.com/SakuraTheBabyfur/Paradise/tree/nanoui_makes_ai_happy

Crazylemon64 commented:
Ah, I'd say it works great! It's easy to spot which cameras are broken, and one can hop to any network in an instant! However, I don't play AI regularly, so perhaps my judgement might not carry much weight... Perhaps, you might want to add a mode to the AI, which determines whether the UI button pulls up your camera monitor or the old camera list?

SakuraTheBabyfur commented:
as i had said in the last post, the UI button now brings up the new camera module. if you wish to use the old camera list, it is available under the AIs "AI commands" tab

Fox-McCloud commented:
@Markolie 

I know next to nothing about the camera system. Mind looking at this?

Markolie commented:
Psi is right - this is a bit of a copy paste. The subsystem should apply to both the camera console and the AI option.

I do think this is a good change, but I personally never run into issues with the security camera console. It could probably do with a touch up though.

Crazylemon64 commented on code/modules/nano/modules/camera_monitor.dm:
```diff
@@ -0,0 +1,170 @@
+	current = null
+
+//Camera control: moving.
+/datum/nano_module/camera_monitor/proc/jump_on_click(var/mob/user,var/A)
```
This doesn't look like this proc is used here...

Crazylemon64 commented on code/modules/nano/modules/camera_monitor.dm:
```diff
@@ -0,0 +1,170 @@
+	A.client.eye = A.eyeobj
+	return 1
+
+/datum/nano_module/camera_monitor/proc/set_current(var/obj/machinery/camera/C)
```
This one does not appear to be used, either

Fox-McCloud commented:
@SakuraTheBabyfur 

Status?

SakuraTheBabyfur commented:
Removed everything that was unneeded, a few functions slipped through the last time I removed stuff. Also made the "SS13" network on by default

Fox-McCloud commented:
I still don't believe this does what @Markolie requested.

Markolie commented:
Yup - this'll also need to apply to the camera console itself. Have a look at the crew monitor for an example on how to do it: https://github.com/ParadiseSS13/Paradise/blob/6deffa0535dab8a895be8a0d57d0fd0891bb4fab/code/game/machinery/computer/crew.dm

SakuraTheBabyfur commented:
@Markolie I'm not entirely sure what you mean by make it apply to the console as well, if you mean the UI part, it already does since they both use the sec_camera.tmpl

Crazylemon64 commented:
I think Mark wanted the camera console to work off of the datum and procs you used in this PR.

SakuraTheBabyfur commented:
I dont think it would work to have the camera console work off of this module, since i removed a lot of checks and stuff and the code assumes AI as the user. I suppose it could be done by checking the user and everything before the proc is called, however im not sure when im going to be able to do it since school is getting very busy lately. Il try to see what i can do but i make no promises, and it probably wont be for another week or two at the earliest

Fox-McCloud commented:
@SakuraTheBabyfur 

Status update?

Fox-McCloud commented:
Closing this due to lack of response+updates.

If you want to tweak this up to @Markolie's specifications, then feel free to re-open this when you have the time.
